### PR TITLE
NullReferenceException in CSharp target #2304   

### PR DESCRIPTION
**Pull request by xsIceman**
updated SingletonPredictionContext in the C# runtime to match that of Java as described in issue #2304. I face the same issue with my IPS parser

### CODE REVIEW

File: runtime/CSharp/runtime/CSharp/Antlr4.Runtime/Atn/SingletonPredictionContext.cs [hunk 1]
```diff
@@ -79,7 +79,7 @@ public override bool Equals(object o)
                 return false;
             }
 			Antlr4.Runtime.Atn.SingletonPredictionContext other = (Antlr4.Runtime.Atn.SingletonPredictionContext)o;
-            return returnState == other.returnState && parent.Equals(other.parent);
+            return returnState == other.returnState && (parent != null && parent.Equals(other.parent));
```
Comment (ericvergnaud) [inline runtime/CSharp/runtime/CSharp/Antlr4.Runtime/Atn/SingletonPredictionContext.cs]: How about when parent is null?

Comment (xsIceman) [reply runtime/CSharp/runtime/CSharp/Antlr4.Runtime/Atn/SingletonPredictionContext.cs]: That is the change i want to add, when parent == null i want to return false. before we did a compare on null and got a nullrefexception. This is the same check that is done in java code SingletonPredictionContext.java.

Comment (ericvergnaud) [reply runtime/CSharp/runtime/CSharp/Antlr4.Runtime/Atn/SingletonPredictionContext.cs]: Then maybe the Java code also needs to be fixed. Please get this right - maybe add a reusable utility to compare objects.

Comment (parrt) [reply runtime/CSharp/runtime/CSharp/Antlr4.Runtime/Atn/SingletonPredictionContext.cs]: It looks like what the Java target does and I'm terrified of changing what that target does so maybe we should allow this within the C# to match the functionality at least?

**Comment by zhech2** [comment]
I am also running into this issue.  For starters can the behavior of the c# code be changed to match the java code. Which this pull request already does?

**Comment by ericvergnaud** [comment]
@parrt blessed

**Comment by parrt** [comment]
@xsIceman can you fix the contributors.txt file? It has weird old date: 2018/10/08,

**Comment by parrt** [comment]
@xsIceman can you fix the contributors.txt file? It has weird old date: 2018/10/08?

**Comment by yonigibbs** [comment]
Hi @parrt, is this PR planned to be merged at some point? Is there anything we can do to help here? We're running into the same issue. Thanks!

**Comment by parrt** [comment]
I guess date is ok. just old.

**Comment by parrt** [comment]
Hmm...seems this [fails to pass a test](https://travis-ci.org/github/antlr/antlr4/jobs/734045778#L2955) @ericvergnaud can you take a peek?

**Comment by ericvergnaud** [comment]
@parrt the failure seems to be with a swift test so not sure how it can relate to this PR. Did you also merge other stuff?

**Comment by parrt** [comment]
I don't think I've included any swift stuff in a while.  maybe I need to try to run the tests locally.  It looks like the npm thing was missing  on an AppVeyor fail.  did you see this one? https://ci.appveyor.com/project/parrt/antlr4/builds/35672295

 maybe it got fixed later

**Comment by ericvergnaud** [comment]
That is unrelated I think.
The npm thing was from me, I disabled JS on AppVeyor for now due to that issue, and asked help from AppVeyor experts but it’s not blocking (the tests pass on Travis)

Envoyé de mon iPhone

**Pull request merged**